swift/apple/FirezoneKit/Sources/FirezoneKit/Stores/Store.swift

```diff
@@ -46,89 +49,110 @@ public final class Store: ObservableObject {
       }
     }
 
+    // Load our state from the system. Based on what's loaded, we may need to ask the user for permission for things.
+    initNotifications()
+    initSystemExtension()
+    initVPNConfiguration()
+  }
+
+  func initNotifications() {
     Task {
-      // Load user's decision whether to allow / disallow notifications
       self.decision = await self.sessionNotification.loadAuthorizationStatus()
+    }
+  }
 
-      // Load VPN configuration and system extension status
-      do {
-        try await self.bindToVPNConfigurationUpdates()
-        let vpnConfigurationStatus = self.status
-
+  func initSystemExtension() {
 #if os(macOS)
-        let systemExtensionStatus = try await self.checkedSystemExtensionStatus()
-
-        if systemExtensionStatus != .installed
-            || vpnConfigurationStatus == .invalid {
-
-          // Show the main Window if VPN permission needs to be granted
-          AppView.WindowDefinition.main.openWindow()
-        } else {
-          AppView.WindowDefinition.main.window()?.close()
-        }
+    Task {
+      do {
+        self.systemExtensionStatus = try await self.checkSystemExtensionStatus()
+      } catch {
+        Log.error(error)
+      }
+    }
 #endif
+  }
 
-        if vpnConfigurationStatus == .disconnected {
-
-          // Try to connect on start
-          try self.vpnConfigurationManager.start()
+  func initVPNConfiguration() {
+    Task {
+      do {
+        // Try to load existing configuration
+        if let neTunnelProviderManager = try await VPNConfigurationManager.loadFromPreferences() {
+          try await setupTunnelObservers(from: neTunnelProviderManager, autoStart: true)
```

Comment:
This could just be:

```swift
vpnConfigurationManager = try await VPNConfigurationManager.loadFromPreferences();
```

Not sure about the session, can it actually be initialised when you've just loaded it from preferences?

Reply:
> Not sure about the session, can it actually be initialised when you've just loaded it from preferences?

Yes, the session (`NETunnelProviderManager.connection`) is available as soon as the manager is loaded.